File: src/main/java/idlezoo/game/domain/Building.java
package idlezoo.game.domain;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import org.immutables.value.Value;

@JsonDeserialize(builder = ImmutableBuilding.Builder.class)
@Value.Immutable
public interface Building {
    // logic start
    default double income(int level) {
        return getBaseIncome() + getBaseIncome() * level * level;
    }

    default double upgradeCost(int level) {
        return getBaseUpgrade() * Math.pow(2, level);
    }

    default double buildCost(int index) {
        // For now taken from Coockie-clicker
        return getBaseCost() * Math.pow(1.15, index);
    }
    // logic end

    double getBaseCost();

    double getBaseIncome();

    double getBaseUpgrade();

    String getName();
}

File: src/main/kotlin/idlezoo/game/domain/Building.kt
package idlezoo.game.domain

import com.fasterxml.jackson.annotation.JsonPropertyOrder

@JsonPropertyOrder(alphabetic = true)
data class Building(val baseCost: Double, val baseIncome: Double, val baseUpgrade: Double, val name: String) {
    fun income(level: Int): Double = baseIncome + baseIncome * level * level
    fun upgradeCost(level: Int): Double = baseUpgrade * Math.pow(2.0, level.toDouble())
    fun buildCost(index: Int): Double = baseCost * Math.pow(1.15, index.toDouble())
}